WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SettingsUsernameChangerFragment.kt

```diff
@@ -0,0 +1,176 @@
+                        "<b>",
+                        username,
+                        "</b>"
+                ), HtmlCompat.FROM_HTML_MODE_LEGACY
+        )
+
+        AlertDialog.Builder(ContextThemeWrapper(activity, style.Calypso_Dialog_Alert)).apply {
+            setTitle(string.settings_username_changer_confirm_dialog_title)
+            setView(layout)
+            setPositiveButton(
+                    string.settings_username_changer_confirm_dialog_positive_action
+            ) { _, _ -> saveUsername(username) }
+            setNegativeButton(android.R.string.cancel, null)
+            create()
+        }.show().also { alertDialog ->
+            // The change username button is disabled at start.
+            val changeUsernameButton = alertDialog.getButton(AlertDialog.BUTTON_POSITIVE).apply {
+                isEnabled = false
+            }
+
+            // Toggles the enabled property of the button based on the correctness of the username being typed.
+            usernameControl.addTextChangedListener(object : TextWatcher {
+                override fun afterTextChanged(editable: Editable?) {
+                    editable?.let {
+                        changeUsernameButton.isEnabled = (it.toString() == username)
+                    }
+                }
+
+                override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
+                }
+
+                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
+                }
+            })
+            usernameControl.requestFocus()
+        }
+    }
+
+    private fun saveUsername(username: String) {
+        showProgress()
+        val payload = PushUsernamePayload(
+                username, KEEP_OLD_SITE_AND_ADDRESS
+        )
+        mDispatcher.dispatch(AccountActionBuilder.newPushUsernameAction(payload))
+    }
+
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    fun onUsernameChanged(event: OnUsernameChanged) {
+        if (event.isError) {
+            AnalyticsTracker.track(SIGNUP_EMAIL_EPILOGUE_UPDATE_USERNAME_FAILED)
+            AppLog.e(
+                    T.API, "SettingsUsernameChangerFragment.onUsernameChanged: " +
+                    event.error.type + " - " + event.error.message
+            )
+            endProgress()
+            showErrorDialog(SpannableStringBuilder(getString(string.signup_epilogue_error_generic)))
+        } else if (event.username != null) {
+            val result = Bundle().apply { putString(RESULT_USERNAME, event.username) }
+            dialogController.confirm(result)
+            AnalyticsTracker.track(Stat.SIGNUP_EMAIL_EPILOGUE_UPDATE_USERNAME_SUCCEEDED)
+        }
+    }
+
+    fun showProgress() {
+        if (progressDialog == null || progressDialog?.window == null || progressDialog?.isShowing == false) {
+            progressDialog = ProgressDialog(context).apply {
+                isIndeterminate = true
+                setCancelable(true)
```

Comment:
I set `Cancelable` to true because I don't want the UI to be blocked. However, I am thinking about how this behavior will work since I haven't been seen any way of canceling requests that have been dispatched. Thoughts welcome :)

Reply:
Unfortunately, I don't have a good answer for this. We don't support request cancellation... I also don't like UI blocking dialogs so I agree keeping the dialog cancelable sounds like the right choice. @osullivanchris Any thoughts?

Reply:
Which part of the flow are we talking about? Is it the dialogue that appears which the spinning loader when the user has chosen a new username? If so, its already in a dialogue so we could add a cancel button in that dialogue in the usual bottom right location for actions?

Reply:
> If so, its already in a dialogue so we could add a cancel button in that dialogue in the usual bottom right location for actions?

I'd rather avoid an explicit "cancel" action, since as @jd-alexander mentioned we can't actually cancel the action. The request was already dispatched to the server so the action will be performed (unless it fails). Our concern is that if we set `cancelable` to `false` the user won't be able to dismiss the dialog in case of a bug or a long timeout period. (btw afaik we don't currently support progress dialogs with actions)

I personally would keep the dialog as is - cancelable without an explicit "cancel" action. In most cases the dialog will disappear within a second anyway and the number of users who will actually cancel it will be imho very low... Another option that comes to my mind is to show a snackbar when the user cancels the dialog - something like "Can't cancel action. Username might have been already updated".